File: backend/internal/shared/kernel/values/recurring.go
// SPDX-License-Identifier: BUSL-1.1
// SPDX-FileCopyrightText: 2026 Gradion

package values

// The monthly reading of an annual recurring figure.
//
// ARR is stored annually because that is how it is negotiated and how it is
// reported, but a reader comparing a subscription against a monthly cost wants
// it monthly, so the product shows both. The division is the entire content of
// this file, and it lives here because it happens twice — once in Go for
// exports and reports, once in TypeScript for the record page — and the two
// have to produce the same number for the same input.
//
// The rounding is HALF UP on the minor unit, and there is no way to divide
// twelve into an arbitrary annual figure without one: EUR 100.00 a year is
// 8.333… a month, and the product has to print something. Half up is chosen
// because it is what a reader doing the division on paper does, and because
// the alternative (banker's rounding) would make two adjacent figures round in
// opposite directions for no reason a reader could see.
//
// The result is APPROXIMATE whenever the division was not exact, and that fact
// travels with it rather than being left for the caller to re-derive. A
// monthly figure printed as exact when twelve of it do not add back to the
// annual one is a figure somebody will eventually multiply and query.

// MonthlyEquivalent divides an annual recurring figure into a monthly one, in
// the same minor units. Approximate reports whether the division lost anything,
// so a surface can mark the figure rather than implying twelve of it reproduce
// the year.
//
// It works on minor units alone and never looks at the currency: the scale is
// already applied on both sides of the division, so dividing minor units by
// twelve is the same operation for a currency with no minor unit as for one
// with three.
//
// A negative input cannot occur — deal_expected_arr_nonnegative and its
// contract twin refuse one in the database, and moneyPairError refuses one
// before that — but the truncation below rounds toward zero, so the half-up
// correction is written to hold for a negative anyway rather than silently
// meaning half-down there.
func MonthlyEquivalent(annualMinor int64) (monthlyMinor int64, approximate bool) {
	const monthsPerYear = 12
	quotient := annualMinor / monthsPerYear
	remainder := annualMinor % monthsPerYear
	if remainder == 0 {
		return quotient, false
	}
	// Half up: a remainder of 6 or more (in absolute terms) carries. Doubling
	// the remainder rather than comparing against 6 keeps the test exact for
	// any divisor, and cannot overflow because the remainder is under 12.
	if remainder < 0 {
		if -remainder*2 >= monthsPerYear {
			quotient--
		}
		return quotient, true
	}
	if remainder*2 >= monthsPerYear {
		quotient++
	}
	return quotient, true
}

File: frontend/src/format/recurring.ts
/**
 * The monthly reading of an annual recurring figure.
 *
 * The twin of `values.MonthlyEquivalent` in the Go kernel, and the two are
 * held together by one corpus rather than by inspection: both this module's
 * test and the Go one read
 * `backend/internal/shared/kernel/values/testdata/monthly_equivalent.json`, so
 * a case answered differently on either side fails there.
 *
 * The rounding is HALF UP on the minor unit, because an arbitrary annual
 * figure does not divide into twelve and the product still has to print
 * something. The result is approximate whenever the division lost anything,
 * and that travels with it: twelve of an approximate monthly figure do not add
 * back to the year, and a surface printing it as exact invites somebody to
 * multiply it.
 */

/** The annual figure divided into months, with whether anything was lost. */
export type MonthlyReading = Readonly<{
  monthlyMinor: number;
  approximate: boolean;
}>;

const MONTHS_PER_YEAR = 12;

/**
 * Divides an annual recurring figure into a monthly one, in the same minor
 * units.
 *
 * Never reads the currency: the minor-unit scale is already applied on both
 * sides of the division, so a currency with no minor unit divides exactly as
 * one with three does.
 *
 * A figure this cannot divide exactly returns null rather than throwing. The
 * server stores ARR as a 64-bit integer and JSON hands it here as a double, so
 * a figure past Number.MAX_SAFE_INTEGER arrives having already lost precision —
 * real, legal, and not something this can read. Throwing would take the whole
 * record page down over a derived line beside a figure that displays fine on
 * its own; the callers render the annual figure and omit the monthly one.
 */
export function monthlyEquivalent(annualMinor: number): MonthlyReading | null {
  if (!Number.isSafeInteger(annualMinor)) {
    return null;
  }
  const quotient = Math.trunc(annualMinor / MONTHS_PER_YEAR);
  const remainder = annualMinor % MONTHS_PER_YEAR;
  if (remainder === 0) {
    return { monthlyMinor: quotient, approximate: false };
  }
  // Half up: a remainder of six or more carries. Doubling rather than
  // comparing against six keeps the test exact and matches the Go side line
  // for line, which is what makes the two readable against each other.
  const carries = Math.abs(remainder) * 2 >= MONTHS_PER_YEAR;
  const adjusted = carries ? quotient + (remainder < 0 ? -1 : 1) : quotient;
  return { monthlyMinor: adjusted, approximate: true };
}
